src/Orleans/Runtime/RequestContext.cs

```diff
@@ -143,36 +148,37 @@ public static void Import(Dictionary<string, object> contextData)
         {
             var values = CallContextData.Value;
 
+#if !NETSTANDARD
             if (PropagateActivityId)
             {
-                var activityId = ActivityId.Value;
-                var activityIdHeader = E2_E_TRACING_ACTIVITY_ID_HEADER;
-#if !NETSTANDARD
-                if (activityId == Guid.Empty)
-                {
-                    activityIdHeader = E2_E_TRACING_LEGACY_ACTIVITY_ID_HEADER;
-                    activityId = Trace.CorrelationManager.ActivityId;
-                }
-#endif
-                
-                if (activityId != Guid.Empty)
+                var activityIdOverride = Trace.CorrelationManager.ActivityId;
+                if (activityIdOverride != Guid.Empty)
                 {
-                    values = values == null ? new Dictionary<string, object>() : new Dictionary<string, object>(values); // Create new copy before mutating data
-                    values[activityIdHeader] = activityId;
-                    // We have some changed data, so write RC data back into LogicalCallContext
-                    CallContextData.Value = values;
+                    object existingActivityId;
+                    if (values == null 
+                        || !values.TryGetValue(E2_E_TRACING_ACTIVITY_ID_HEADER, out existingActivityId)
+                        || activityIdOverride != (Guid)existingActivityId)
+                    {
+                        // Create new copy before mutating data
+                        values = values == null ? new Dictionary<string, object>() : new Dictionary<string, object>(values);
+                        values[E2_E_TRACING_ACTIVITY_ID_HEADER] = activityIdOverride;
+                    }
                 }
-               
             }
-            if (values != null && values.Count != 0)
-                return (Dictionary<string, object>)serializationManager.DeepCopy(values);
-            return null;
+#endif
+
+            return (values != null && values.Count > 0)
+                ? (Dictionary<string, object>)serializationManager.DeepCopy(values)
```

Comment:
In case of non-empty ActivityId this will clone the dictionary once again (first clone is at line 163).